src/back/spv/selection.rs

```diff
@@ -0,0 +1,250 @@
+//! Generate SPIR-V conditional structures.
+//!
+//! Builders for `if` structures with `and`s.
+//!
+//! The types in this module track the information needed to emit SPIR-V code
+//! for complex conditional structures, like those whose conditions involve
+//! short-circuiting 'and' and 'or' structures. These track labels and can emit
+//! `OpPhi` instructions to merge values produced along different paths.
+//!
+//! This currently only supports exactly the forms Naga uses, so it doesn't
+//! support `or` or `else`, and only supports zero or one merged values.
+//!
+//! Naga needs to emit code roughly like this:
+//!
+//! ```ignore
+//!
+//!     value = DEFAULT;
+//!     if COND1 && COND2 {
+//!         value = THEN_VALUE;
+//!     }
+//!     // use value
+//!
+//! ```
+//!
+//! Assuming `ctx` and `block` are a mutable references to a [`BlockContext`]
+//! and the current [`Block`], and `merge_type` is the SPIR-V type for the
+//! merged value `value`, we can build SPIR-V for the code above like so:
+//!
+//! ```ignore
+//!
+//!     let cond = Selection::start(ctx, block, merge_type);
+//!         // ... compute `cond1` ...
+//!     cond.if_true(ctx, block, cond1, DEFAULT);
+//!         // ... compute `cond2` ...
+//!     cond.if_true(ctx, block, cond2, DEFAULT);
+//!         // ... compute THEN_VALUE
+//!     let merged_value = cond.finish(ctx, block, THEN_VALUE);
+//!
+//! ```
+//!
+//! After this, `merged_value` is either `DEFAULT` or `THEN_VALUE`, depending on
+//! the path by which the merged block was reached.
+//!
+//! This takes care of writing all branch instructions, including an
+//! `OpSelectionMerge` annotation in the header block; starting new blocks and
+//! assigning them labels; and emitting the `OpPhi` that gathers together the
+//! right sources for the merged values, for every path through the selection
+//! construct.
+//!
+//! When there is no merged value to produce, you can pass `()` for `merge_type`
+//! and the merge values. In this case the `finish` method returns `()`.
+
+#![allow(dead_code)] // until texture bounds checks
+
+use super::{Block, BlockContext, Instruction};
+use spirv::Word;
+
+/// A private struct recording what we know about the selection construct so far.
+pub(super) struct Selection<M: MergeTuple> {
```

Comment:
In terms of Valid Usage, is it pretty much required that the same `&mut Block` is passed to all of the methods here? If that's true, maybe it makes sense to just borrow it mutably for the duration of `Selection` lifetime?

Reply:
That's a good idea. Valid Usage requires using this in a properly nested way with respect to other control structures, and having `Selection` take ownership of the `&mut Block` would help enforce that. But in between calls to `start`, `if_true`, and `finish`, we need that `&mut Block` for writing the body instructions of the conditional. So `Selection` would need a method for reborrowing the block out of it.

This suggestion also made me realize that `Selection::current_label` would go stale if somebody switched blocks between calls to `Selection` methods - which is certainly necessary in some cases. As it turns out, `current_label` is unnecessary, so I took it out.

Reply:
This is implemented in the latest push. Let me know what you think.